### DISCUSSION TITLE
Multi-tenancy, proprietary server code, and other considerations.

### DISCUSSION THREAD

**Discussion by leonpolak** [metarhia/Roadmap#9] [General]
Just to write down some discussions we previously have had and to be sure everyone understands the logic behind the critical parts of the roadmap and their priorities.

1) Multi-tenancy

There are many types of businesses, especially SaaS businesses, that need to separate access to source code and customizable static content between different subscribers. For example, hosting providers, marketplaces, etc. Even if the business that installs Metarhia does not require access to the IDE for subscribers, but only uses it itself, default multi-tenancy will be also handy, for example, for AB testing.

Separating the front-end code seems easy by putting the tenant code in a "static/tenants" subfolder and redirecting to it via subdomains, URL path parts, or whatever. However, we currently have a problem with the server code. There are 4 types of files in the system that we allow editing: schemas, API (RPC) methods, flows (processes), and flow steps (procedures). And there is no strict isolation between system schemas and apis, and tenant schemas and apis. As well, as between files of different tenants. In my opinion, there could be a special "tenants" folder on the "server side" outside of the "static" folder, where we can concentrate all the tenants code in subfolders. 

For example, the flow step files should be in the "tenants/tenantId/procedures/store" folder, not in the "store" folder as it is now. They can be organized in different ways for the efficiency of the system, but we need to have clear identifiers for what to show tenants and how to manage those files.

We have the same problem with database isolation. I think we could define in the server config file the number of databases we should preallocate to tenants and then grant access at registration (with batch reallocation if necessary).

2) Proprietary "server-side" code

We should prepare the system for proprietary code parts, keeping the open-source core intact and easily synchronized with the master branch. I suggest we create a default "application/extensions" folder excluded from git with a default dependencies.js file inside, allowing to develop custom extensions (for example serving particular industries). The main dependencies.js  file importing extensions dependencies will allow tenant apis access to proprietary modules. I'm not sure how to include sub files to process.json, but we should find a neat way to separate open-source and proprietary parts, otherwise, working on closed branches will result in constant conflicts with the master.

3) Motivation

So far I don't see special enthusiasm in the majority of participants, but I understand multiple reasons against active contribution. Full time jobs, families, unfamiliar ecosystem, overall code complexity, high standards, etc. However, the main reason IMHO is unclear personal benefits and objectives. I as well don't share the joy of spending a lot of time just to show how good you are in open-source, or to make the world better.  

But, here is the trick: we don't build yet another soon forgotten library, we build the unique powerful universal fundament for a lot of really profitable projects to come. Either you by yourself will find a business opportunity to make a startup on top of this fundament, or you will join someone else's project, your current efforts will not be in vain. At least I anticipate unusually high wages and relocation variants for those, who contribute at the early stages and thus obtain profound knowledge of the ecosystem. I have no slightest doubt about it.

**Comment by leonpolak** [comment]
4) Postgres table for tenant external APIs info

We need an additional table to store tenants' credentials for external APIs access, for example, GMail Oauth2. I think we need only the following fields there: 
apiId bigint, 
TenantId bigint 
ApiName varchar 
ApiTipe enum 
ApiData: json

Api Credentials Data is really variable, so, we need to validate CRUD on the application level. Moreover, we will integrate a lot of new APIs incrementally, so, I don't see an elegant way to store it except in jsons. I will be glad to hear other opinions.

5) Contribution ways (motivation application)

Choose a topic from the roadmap, preferably you have some experience with or would like to have in your possession in the future. 
Discuss ways to do it. 
Find some time to work
Make it as a maximally independent module
Complete it (sic!)

**Comment by tshemsedinov** [comment]
1. Multi-tenancy will be repository-based and thread-based, not folder-based and not closure-based. We will use closure-based isolation for simple multi-user system, without custom schemas, apis and so on. Thus we are building secure and reliable app server. Mixing data structures in RAM and files at the disk is not acceptable for us.
2. Each tenant will have separate repo to hold the code (you call "tenant folders") and one or more threads to run this code. Thus we will bill him for at least a single thread, but he can configure thread limit we will dynamically increase in runtime.
3. What we need to do next:
3.1. Implement add/remove tenants without stop of app server
3.2. Increase/decrease thread for each tenant (now it is a constant) without stop of app server
4. Db connections and schemas will be absolutely separated at the level of database.
5. What we share: scheduler, watcher, logger, health monitoring, etc.
